gecoscc/models.py

```diff
@@ -395,23 +400,245 @@ class AdminUserOUManage(colander.MappingSchema):
+    ''' Subclass for validating numeric range of scripts (00-99)
+    
+    Attributes:
+      err_msg (str):    error message
+      pattern (str):    regex for valid numeric range 
+    '''
+    pattern = '^[0-9][0-9]-.*'
+    err_msg = _('Any script out of range (00-99)')
+
+    def __call__(self, node, value):
+
+        super(UpdateScriptRangeValidator, self).__call__(node,value)
+
+        scriptdir = self.decompress + 'scripts'
+ 
+        if os.path.isdir(scriptdir):
+            for script in os.listdir(scriptdir):
+                if not (re.match(self.pattern, script)):
+                    node.raise_invalid(gettext(self.err_msg))
+        
+
+class UpdateControlFileValidator(UpdateBaseValidator):
+    ''' Subclass for verifying requisites of control file
+    
+    Attributes:
+      err_msg (str):    error message
+    '''
+    err_msg = 'Control file requirements not met'
+
+    def __call__(self, node, value):
+
+        from iscompatible import iscompatible, string_to_tuple
+        super(UpdateControlFileValidator, self).__call__(node,value)
+
+        request = pyramid.threadlocal.get_current_request()
+        controlfile = self.decompress + os.sep + 'control'
+
+        settings = get_current_registry().settings
+        api = get_chef_api(settings, request.user)
+        cookbook = get_cookbook(api, settings.get('chef.cookbook_name'))
+
+        if os.path.exists(controlfile):
+            gecoscc_require = cookbook_require = None
+            with open(controlfile,'r') as f:
+                for line in f:
+                    if line.startswith('gecoscc'):
+                        gecoscc_require = line
+                    elif line.startswith('cookbook'):
+                        cookbook_require = line
+                      
+            if gecoscc_require and not iscompatible(gecoscc_require, string_to_tuple(request.VERSION)):
+                node.raise_invalid(gettext(self.err_msg))
+  
+            if cookbook_require and not iscompatible(cookbook_require, string_to_tuple(cookbook['version'])):
+                node.raise_invalid(gettext(self.err_msg))
+         
+
+# Update preparer
+def unzip_preparer(value):
+    '''
+    From deform documentation: "The preparer of a schema node is called after deserialization but before validation."
+    The data is prepared by decompressing the zip file in a temporary directory. After that, validators are called.
+    '''
+    if value is not colander.null:
+        try:
+            if 'fp' in value:
+                # local_file
+                with open('/tmp/' + value['filename'], 'wb') as zipped:
+                    zipped.write(value['fp'].read())
+            else: 
+                # remote_file
+                f = urllib2.urlopen(value['url'])
+                with open('/tmp/' + os.path.basename(value['url']), "wb") as zipped:
+                    zipped.write(f.read())
+
+            # Decompress zipfile into temporal dir
+            tmpdir = tempfile.mkdtemp()
```

Comment:
I've created this temporary directory instead of `updates.tmp` because it is used only during validation of update file. Plus, it is not secure decompress all the uploaded files in the same directory in case some of them fails.